clients/include/testing_asum_strided_batched.hpp

```diff
@@ -0,0 +1,239 @@
+/* ************************************************************************
+ * Copyright 2018-2019 Advanced Micro Devices, Inc.
+ * ************************************************************************ */
+
+#include "cblas_interface.hpp"
+#include "near.hpp"
+#include "rocblas.hpp"
+#include "rocblas_init.hpp"
+#include "rocblas_math.hpp"
+#include "rocblas_random.hpp"
+#include "rocblas_test.hpp"
+#include "rocblas_vector.hpp"
+#include "unit.hpp"
+#include "utility.hpp"
+
+template <typename T1, typename T2 = T1>
+void testing_asum_strided_batched_bad_arg_template(const Arguments& arg)
+{
+    rocblas_int N           = 100;
+    rocblas_int incx        = 1;
+    rocblas_int stridex     = 1;
+    rocblas_int batch_count = 5;
+
+    static const size_t safe_size        = 100;
+    T2                  rocblas_result   = 10;
+    T2*                 h_rocblas_result = &rocblas_result;
+
+    rocblas_local_handle handle;
+    device_vector<T1>    dx(safe_size);
+    if(!dx)
+    {
+        CHECK_HIP_ERROR(hipErrorOutOfMemory);
+        return;
+    }
+
+    CHECK_ROCBLAS_ERROR(rocblas_set_pointer_mode(handle, rocblas_pointer_mode_host));
+
+    EXPECT_ROCBLAS_STATUS((rocblas_asum_strided_batched<T1, T2>(
+                              handle, N, nullptr, incx, stridex, h_rocblas_result, batch_count)),
+                          rocblas_status_invalid_pointer);
+    EXPECT_ROCBLAS_STATUS(
+        (rocblas_asum_strided_batched<T1, T2>(handle, N, dx, incx, stridex, nullptr, batch_count)),
+        rocblas_status_invalid_pointer);
+    EXPECT_ROCBLAS_STATUS((rocblas_asum_strided_batched<T1, T2>(
+                              nullptr, N, dx, incx, stridex, h_rocblas_result, batch_count)),
+                          rocblas_status_invalid_handle);
+}
+
+template <typename T1, typename T2 = T1>
+void testing_asum_strided_batched_template(const Arguments& arg)
+{
+    rocblas_int N           = arg.N;
+    rocblas_int incx        = arg.incx;
+    rocblas_int stridex     = arg.stride_x;
+    rocblas_int batch_count = arg.batch_count;
+
+    double rocblas_error_1;
+    double rocblas_error_2;
+
+    rocblas_local_handle handle;
+
+    rocblas_int abs_incx = incx >= 0 ? incx : -incx;
```

Comment:
Yeah, `asum` does not have a second vector argument, so if `n <= 0` or `incx <= 0`, it's considered empty (return value 0). There's no need to compute `abs_incx`, since `incx > 0` is assumed, after the error-checks.
```
 sasum = 0.0e0
 stemp = 0.0e0
 IF (n.LE.0 .OR. incx.LE.0) RETURN
```

Reply:
> Since we're returning invalid_size if incx < 0, I don't think we need this variable.

BTW, `asum` and  `nrm2` should not return `invalid` if `n <= 0` or `incx <= 0`. They should return `0`.

Reply:
Are we sure we want to match that fortran code and cublas?  The old docs you sent say incx 0 is invalid, and also for single input functions incx must be > 0.       I get the argument for N but as incx can sometimes be negative so returning 0 may make you miss the parameter error, is improving the API an option?

Reply:
We should match the original BLAS specification (that it is in Fortran is incidental).

When vectors are used with other vectors or matrices, the vectors may have negative increments, meaning the vector is read from the end. This generally only involves adjusting the pointer to the vector to the end, while leaving the negative increment alone.

When vectors are used alone (the only non-scalar argument), then if they have a zero or negative size or increment, they behave as if they are zero-length, and the mathematical definition of the operation on a zero-length vector is used.

Matrix dimensions cannot have negative sizes; they are flagged as invalid. However, zero-sized matrix dimensions are usually allowed as long as other constraints are met.